vault/generate_root.go

```diff
@@ -263,29 +264,70 @@ func (c *Core) GenerateRootUpdate(ctx context.Context, key []byte, nonce string,
 		}, nil
 	}
 
-	// Recover the master key
-	var masterKey []byte
+	// Combine the key parts
+	var combinedKey []byte
 	if config.SecretThreshold == 1 {
-		masterKey = c.generateRootProgress[0]
+		combinedKey = c.generateRootProgress[0]
 		c.generateRootProgress = nil
 	} else {
-		masterKey, err = shamir.Combine(c.generateRootProgress)
+		combinedKey, err = shamir.Combine(c.generateRootProgress)
 		c.generateRootProgress = nil
 		if err != nil {
 			return nil, errwrap.Wrapf("failed to compute master key: {{err}}", err)
 		}
 	}
 
-	// Verify the master key
-	if c.seal.RecoveryKeySupported() {
-		if err := c.seal.VerifyRecoveryKey(ctx, masterKey); err != nil {
+	switch {
+	case c.seal.RecoveryKeySupported():
+		// Ensure that the combined recovery key is valid
+		if err := c.seal.VerifyRecoveryKey(ctx, combinedKey); err != nil {
 			c.logger.Error("root generation aborted, recovery key verification failed", "error", err)
 			return nil, err
 		}
-	} else {
-		if err := c.barrier.VerifyMaster(masterKey); err != nil {
-			c.logger.Error("root generation aborted, master key verification failed", "error", err)
-			return nil, err
+		// If we are in recovery mode, then retrieve
+		// the stored keys and unseal the barrier
+		if c.recoveryMode {
```

Comment:
Could the recovery mode unseal logic be moved into the strategy? Seems cleaner if we just pass Core into generateRecoveryToken. WDYT?

Reply:
I think we can do this. I agree that it might be cleaner. While thinking over this, I just realized that we are not considering a case where the recovery token is generated again. We try to unseal the barrier every time we generate a recovery token. I'm not sure what the behavior will be, but I'll need to check that out.

Reply:
Having second thoughts on this now. Currently, we verify that the combined key is able to unseal the barrier and in the case of recovery keys, we fetch the master key and try unsealing the barrier. This is the authentication step. Calling strategy.generate() is the step that follows after a successful authentication. I am feeling like, if we combine those, `strategy.generate` will be responsible for both authentication and token creation. In case of error, we won't know if the error is due to auth failure or something else. Leaning towards thinking that it is cleaner the way it is right now. Let me know your thoughts.

Reply:
Decided over the zoom call that some sort of refactoring is required here to make this cleaner. But it is not imminent and will be done separately post merge, given the time constraints.